bootloader/src/pyi_archive.c

```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+        return -1;
+    }
+
+    /* Ensure endianess of file matches */
+    if (head[0x05] != *(uint8_t*)&endianess_bytes) {
+        VS("LOADER: %s endianess does not match CPU\n", status->archivename);
+        return -1;
+    }
+
+    /* Ensure ELF version is supported */
+    if (head[0x06] != 1 || *(uint32_t*)(head + 0x14) != 1) {
+        VS("LOADER: %s has an unknown ELF version\n", status->archivename);
+        return -1;
+    }
+
+    /* Load ELF section header parameters */
+    bits = head[0x04]; /* 2=64, 1=32 */
+    if (bits == 2) {
+        e_shoff = *(uint64_t*)(head + 0x28);
+        e_shentsize = *(uint16_t*)(head + 0x3a);
+        e_shnum = *(uint16_t*)(head + 0x3c);
+        e_shstrndx = *(uint16_t*)(head + 0x3e);
+    } else {
+        e_shoff = *(uint32_t*)(head + 0x20);
+        e_shentsize = *(uint16_t*)(head + 0x2e);
+        e_shnum = *(uint16_t*)(head + 0x30);
+        e_shstrndx = *(uint16_t*)(head + 0x32);
+    }
+
+    section = malloc(e_shentsize);
+    if (section == NULL) {
+        FATAL_PERROR("malloc", "Could not allocate buffer for section header.");
+        return -1;
+    }
+
+    /* Load shstr section into buffer */
+    fseek(status->fp, e_shoff + e_shentsize * e_shstrndx, SEEK_SET);
+    if (fread(section, e_shentsize, 1, status->fp) != 1) {
+        free(section);
+        VS("LOADER: %s could not read shstr section header\n", status->archivename);
+        return -1;
+    }
+
+    if (bits == 2) {
+        shstr_offset = *(uint64_t*)(section + 0x18);
+        shstr_size = *(uint64_t*)(section + 0x20);
+    } else {
+        shstr_offset = *(uint32_t*)(section + 0x10);
+        shstr_size = *(uint32_t*)(section + 0x14);
+    }
+
+    shstr = malloc(shstr_size);
+    if (shstr == NULL) {
+        free(section);
+        FATAL_PERROR("malloc", "Could not allocate buffer for ELF shstrs.");
+        return -1;
+    }
+
+    fseek(status->fp, shstr_offset, SEEK_SET);
+    if (fread(shstr, shstr_size, 1, status->fp) != 1) {
+        free(section);
+        free(shstr);
+        VS("LOADER: %s could not read shstr section\n", status->archivename);
+        return -1;
+    }
+
+    /* Find section named "pydata" */
+    for (section_idx=0; section_idx<e_shnum; section_idx++) {
+        uint32_t name_offset;
+
+        fseek(status->fp, e_shoff + e_shentsize * section_idx, SEEK_SET);
+        if (fread(section, e_shentsize, 1, status->fp) != 1) {
+            free(section);
+            free(shstr);
+            VS("LOADER: %s could not read section header\n", status->archivename);
+            return -1;
+        }
+
+        name_offset = *(uint32_t*)(section);
+        if (!strncmp(shstr + name_offset, "pydata", shstr_size-name_offset)) {
+            break;
+        }
+    }
+
+    if (section_idx == e_shnum) {
+        free(section);
+        free(shstr);
+        VS("LOADER: %s could not find pydata section\n", status->archivename);
+        return -1;
```

Comment:
This code in repeated quite often. What dou you think about using some (ugly) `goto fail` and having this code only once? Please have a look how this is done in the other parts of the bootloader and implement it the same way.